### PR TITLE
Lincheck guide

### PR DESCRIPTION
**Pull request by mvicsokolova**

### CODE REVIEW

File: docs/lincheck-test-tutorial.md [hunk 1]
```diff
@@ -32,17 +33,20 @@ repositories {
...
    testImplementation "org.jetbrains.kotlin:kotlin-test"
 }
 ```
 
-## Create a test
+## Write your first Lincheck test
  
 Create `CounterTest.kt` file in the directory `src/test/kotlin` and put the following code there:
 
+TODO: should we include imports in the code blocks?
```
Comment (mvicsokolova) [inline docs/lincheck-test-tutorial.md:47]: Yes, I will add imports here. We will later omit imports (and any distractive code) in the runnable samples on the site using publication plugin, like this:

```
import org.jetbrains.kotlinx.lincheck.annotations.Operation
import org.jetbrains.kotlinx.lincheck.check
import org.jetbrains.kotlinx.lincheck.strategy.stress.StressOptions
import org.jetbrains.kotlinx.lincheck.verifier.VerifierState
import org.junit.Test

//sampleStart
class CounterTest {
   private val c = Counter() // initial state

   // operations on the Counter
   @Operation fun inc() = c.inc()
   @Operation fun get() = c.get()

   @Test // JUnit
   fun test() = StressOptions().check(this::class) // the magic button
}
//sampleEnd
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
```

And the runnable sample will contain code between  `//sampleStart` `//sampleEnd` only.

File: docs/lincheck-test-tutorial.md [hunk 2]
```diff
@@ -91,41 +97,38 @@ class CounterTest {
...
-4. The execution switched back to the Thread 2
-5. Thread 2 increments the old value of the counter that was read before the switch and returned `1`
-
+In this trace, the following list of events have occurred:  
+**T2:** The 2nd thread reads the current value of the counter (`value.READ: 0`) and pauses.    
+**T1:** The 1st thread executes `inc()`, which returns `1`, and finishes.  
+**T2:** The 2nd thread increments the previously read value of the counter and incorrectly updates it to `1`, returning `1` as a result.
 
 ## Test structure
+TODO: I am not sure that we should discuss the test structure in the "Lincheck Hello World" section.
```
Comment (mvicsokolova) [inline docs/lincheck-test-tutorial.md:131]: Maybe, you're right, no need to overload the intro. I'll move the explanation of the structure to the next section then

File: docs/testing-modes.md [hunk 3]
```diff
@@ -1,275 +1,260 @@
...
-In this section we will take a closer look at the testing modes provided by Lincheck.
+In this section we will take a closer look at the testing modes provided by Lincheck. 
 
-TODO: why do you need stack as an example here? Why counter from the previous example is worse?
-TODO: It seems that you focus on the test structure, not on the testing modes. Let's re-think the guide structure.
-
-For an example we will test the lock-free Treiber stack algorithm. 
-In addition to the standard `push(value)` and `pop()` operations, we have also implemented a non-linearizable (thus, incorrect) `size()`, 
-which increases and decreases the corresponding `size` field following successful push and pop invocations.
+We are going to explain the structure of the test and what automatically happens under the hood for both stress and
```
Comment (mvicsokolova) [inline docs/testing-modes.md:5]: Now with the Counter example from the previous section it looks better to me. The Stack implementation was indeed just an extra distraction and was not used in any way to explain the testing modes.

File: docs/testing-modes.md [hunk 4]
```diff
@@ -1,275 +1,260 @@
...
-The stress test revealed the incorrect behaviour of `size()` implementation: negative return value. 
-Here we come again to the main advantage of the model checking mode: providing a trace to reproduce the error.
+Then `Lincheck` starts real threads, synchronizes them in an active manner to guarantee that the operations are performed in parallel, 
+and executes the operations, repeating the run many times in the hope to hit an interleaving that produces an incorrect result. 
+
+To increase the number of different interleavings, we also add random busy-wait cycles between the operations and inside them. 
+
+Below is the high-level picture of how one of the stress executions for the `Counter` may be performed.
+
+![Stress execution of the Counter](../docs/images/counter_stress_mode.png)
```
Comment (mvicsokolova) [inline docs/testing-modes.md:63]: I think the picture of stress testing execution may be quite useful for getting the idea.

Though we should do something about it's size

File: docs/progress-guarantees.md [hunk 5]
```diff
@@ -61,37 +62,32 @@ class DataHolderTest {
 }
 ```
 
-TODO: I would suggest adding a "To sum up" subsection to each part, and adding some words on what we have learned along with the full code. A link to the next chapter can also be put there.
-
-> Get the full code [here](../src/jvm/test/org/jetbrains/kotlinx/lincheck/test/guide/DataHolderTest.kt).
-
 The test will fail with the following output:
 
 TODO: but an active lock has been DETECTED, let's make a PR
```
Comment (mvicsokolova) [inline docs/progress-guarantees.md:67]: I didn't get you, please explain what is wrong with this output?

File: docs/verification.md [hunk 6]
```diff
@@ -27,6 +27,10 @@ How to define a sequential specification:
 
 As an example, let's define the sequential specification for the Multi-Producer-Single-Consumer queue (see the [section about execution constraints](constraints.md) for the full code).
 Sequential specification for the concurrent `MpscLinkedAtomicQueue` will delegate it's operations to the `java.util.LinkedList<Int>`.
+
+TODO: is possible to replace
```
Comment (mvicsokolova) [inline docs/verification.md:31]: Could you also please explain this TODO?

**Review by ndkoval — CHANGES_REQUESTED** [review]
@SokolovaMaria Great tutorial! I did a pass over all sections (except for the verification one), fixed some bugs, and paraphrased some parts. Some issues are still unresolved, I added the corresponding comments, please go through them. Please be so kind to review my changes as well.

File: docs/topics/testing-strategies.md [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+class Counter {
+    @Volatile
+    private var value = 0
+
+    fun inc(): Int = ++value
+    fun get() = value
+}
+```
+
+## Stress testing
```
Comment (ndkoval) [inline docs/topics/testing-strategies.md:17]: We need an example with "happens-before" clocks:
```
values in "[..]" brackets indicate the number of completed operations 
in each of the parallel threads seen at the beginning of the current operation
```

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: docs/lincheck-test-tutorial.md (hunk 1 above)
Comment (ndkoval) [reply docs/lincheck-test-tutorial.md]: Could you please insert //sampleStart and //sampleEnd everywhere?

File: docs/progress-guarantees.md (hunk 5 above)
Comment (ndkoval) [reply docs/progress-guarantees.md]: "an active lock has been detected" sounds more natural than "an active lock has been found". Anyway, let's forget about this.

File: docs/topics/operation-arguments.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+   
+       @Operation
+       fun get(@Param(name = "key") key: Int) = map.get(key)
+   
+       @Test
+       fun stressTest() = StressOptions().check(this::class)
+   }
+   ```
+
+5. Run the `stressTest()` and see the following output:
```
[the remark this answers is not in the document]
Comment (mvicsokolova) [reply docs/topics/operation-arguments.md:83]: The `modelCheckingTest` for this example with MultiMap will take a lot of time and probably fail with OutOfMemory error. This case is covered in the next section about modular testing.

File: docs/topics/progress-guarantees.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+Lincheck has found the following blocking interleaving:
+
+1. **T1:** The first thread (the writer) increments the version and switches its execution 
+   to the second thread before updating the data.
+2. **T2:** The second thread (the reader) reads the `version == 1` and 
+   starts to spin in a loop, waiting for the concurrent `update(..)` to finish.
+
+If the first thread hangs, the second thread stays in an active loop with no progress. 
+
+The common non-blocking progress guarantees are (in order to weaken):
```
[the remark this answers is not in the document]
Comment (mvicsokolova) [reply docs/topics/progress-guarantees.md:139]: @danil-pavlov could you please tell what is the best way to highlight this piece of theoretical information? 🙏🏻

File: docs/kl.tree [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        SYSTEM "https://resources.jetbrains.com/stardust/product-profile.dtd">
+
+<product-profile id="kl"
+                 name="kotlinx-lincheck"
+                 start-page="lincheck-guide.md">
+
+    <chunk include-id="lincheck">
+        <toc-element id="lincheck-guide.md"/>
+        <toc-element id="introduction.md"/>
+        <toc-element id="testing-strategies.md" toc-title="Testing strategies"/>
```
[the remark this answers is not in the document]
Comment (mvicsokolova) [reply docs/kl.tree:12]: @danil-pavlov could you please elaborate on this?)

File: README.md [hunk 11]
```diff
@@ -16,8 +16,8 @@ Please read [the tutorial](/docs/topics/lincheck-guide.md) that explains how to
 ##### Given talks:
  *  [Lincheck. Testing concurrent data structures in Java](https://www.youtube.com/watch?v=YAb7YoEd6mM) (Heisenbug 2019, RU)
  *  [Testing concurrent algorithms with Lincheck](https://nkoval.com/talks/#lincheck-joker-2019) (Joker 2019, RU)
```
Comment (mvicsokolova) [inline README.md:18]: @ndkoval could you please provide the link for the corresponding video, I could not find it. And also don't forget to update the same reference in `Lincheck-guide.md`.

File: docs/topics/blocking-data-structures.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
+[//]: # (title: Testing blocking data structures)
+
+Lincheck supports testing blocking data structures implemented with [suspending functions](https://github.com/Kotlin/KEEP/blob/master/proposals/coroutines.md#coroutines-overview)
+from the Kotlin language. The examples of such data structures from the `kotlinx.coroutines` library are mutexes, semaphores, and channels.
+
+For more information on these data structures, see the [Coroutines guide](https://kotlinlang.org/docs/reference/coroutines/coroutines-guide.html).
+
+### Dual data structures
```
[the remark this answers is not in the document]
Comment (mvicsokolova) [reply docs/topics/blocking-data-structures.md]: Done, removed the section about blocking data structures for now with all the references, I hope

File: docs/lincheck-test-tutorial.md (hunk 1 above)
Comment (danil-pavlov) [reply docs/lincheck-test-tutorial.md]: //sampleStart and //sampleEnd works for hiding code lines only in runnable code snippets. This snippet cannot be runnable, it doesn't have an output

File: docs/kl.tree (hunk 10 above)
[the remark this answers is not in the document]
Comment (danil-pavlov) [reply docs/kl.tree:12]: "toc-title" helps to create an alternative title for the left-hand menu (document tree). It's useful when the actual page title is long, and you need a shorter version for the menu

All other pages use the same title for the menu, but this particular one was too long

File: docs/lincheck-test-tutorial.md (hunk 1 above)
Comment (ndkoval) [reply docs/lincheck-test-tutorial.md]: @danil-pavlov I expected this to be a runnable code snippet as it is a JUnit test, we have a ton of similar snippets all over the guide. What is the best way to make them runnable?

File: README.md (hunk 11 above)
Comment (ndkoval) [reply README.md:18]: https://www.youtube.com/watch?v=_0_HOnTSS0E

File: docs/kl.tree (hunk 10 above)
[the remark this answers is not in the document]
Comment (ndkoval) [reply docs/kl.tree:12]: @danil-pavlov thanks for the clarification!

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyGuarantee.kt [hunk 13]
```diff
@@ -21,11 +21,22 @@
...
  * ManagedGuarantee will be constructed for classes with a name listed in [fullClassNames].
  */
 fun forClasses(vararg fullClassNames: String) = forClasses { it in fullClassNames }
 
+/**
+ * ManagedGuarantee will be constructed for all [classes].
+ */
+fun forClasses(vararg classes: KClass<*>) = forClasses {
+    it in classes.map { kClass ->
+        kClass.qualifiedName ?: error("The class $it is local or a it is class of an anonymous object")
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyGuarantee.kt:36]: "or a it is class of an anonymous object" -> "or is a class of an .. "

File: docs/topics/operation-arguments.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ndkoval) [reply docs/topics/operation-arguments.md:83]: It should work, please check locally. Otherwise, our model checking is absolutely useless 😄

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (mvicsokolova) [reply README.md:18]: I guess this talk refers to: Lock-free algorithms testing (Joker 2017, RU) 🙂

File: docs/lincheck-test-tutorial.md (hunk 1 above)
Comment (danil-pavlov) [reply docs/lincheck-test-tutorial.md]: We'll look into it with the website team. Probably, we can find a solution

File: docs/topics/operation-arguments.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+   ```
+
+Due to the small range of keys, Lincheck quickly revealed the race bug: when two values are being added concurrently by the same key, 
+one of the values may be overwritten and lost.
+
+> See the source code [here](https://github.com/Kotlin/kotlinx-lincheck/blob/guide/src/jvm/test/org/jetbrains/kotlinx/lincheck/test/guide/MultiMapTest.kt).
+>
+{type="note"}
+
+## What's next
```
[the remark this answers is not in the document]
Comment (danil-pavlov) [reply docs/topics/operation-arguments.md:124]: I'll add the link during the final editing

File: README.md (hunk 11 above)
Comment (ndkoval) [reply README.md:18]: Fixed

**Review by ndkoval — APPROVED** [review]
@mvicsokolova the guide looks great now! Let's ask somebody to proofread it and publish the guide after that.

File: README.md [hunk 15]
```diff
@@ -10,11 +10,14 @@ The artifacts are available in [Bintray](https://bintray.com/kotlin/kotlinx/kotl
 Use `org.jetbrains.kotlinx:lincheck:<version>` artifact path in Gradle 
 and `org.jetbrains.kotlinx:lincheck-jvm:<version>` in Maven.
 
+##### Guide:
+Please read [the tutorial](/docs/topics/lincheck-guide.md) that explains how to use Lincheck and presents most of the features step-by-step.
```
Comment (koshachy) [inline README.md:14]:
```suggestion
[Read the tutorial](/docs/topics/lincheck-guide.md) that explains how to use Lincheck and presents most of the features step-by-step.
```

File: docs/topics/lincheck-guide.md [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+concurrent Lincheck test contains only about 15 lines.
+
+When given a list of operations, Lincheck automatically:
+
+* Generates a set of random concurrent scenarios.
+* Examines them using either stress-testing or bounded model checking.
+* Verifies that the results of each invocation satisfy the required correctness property (linearizability is the default
+one).
+
+This guide will help you get in touch with the framework and try the most useful features with examples. Explore the
```
Comment (koshachy) [inline docs/topics/lincheck-guide.md:17]: Maybe a separate section for this one?

File: docs/topics/introduction.md [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+   repositories {
+       mavenCentral()
+   }
+   
+   dependencies {
+       // Lincheck dependency
+       testImplementation("org.jetbrains.kotlinx:lincheck:2.14.1")
+   
+       // This dependency allows you to work with kotlin.test and JUnit:
+       testImplementation("junit:junit:4.13")
```
Comment (koshachy) [inline docs/topics/introduction.md:44]:
```suggestion
       // This dependency allows you to work with kotlin.test and JUnit:
       testImplementation("junit:junit:4.13")
```

File: docs/topics/introduction.md [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+   
+   ```groovy
+   repositories {
+       mavenCentral()
+   }
+   
+   dependencies {
+       // Lincheck dependency
+       testImplementation "org.jetbrains.kotlinx:lincheck:2.14.1"
+       // This dependency allows you to work with kotlin.test and JUnit
```
Comment (koshachy) [inline docs/topics/introduction.md:59]:
```suggestion
       // This dependency allows you to work with kotlin.test and JUnit:
```

File: docs/topics/introduction.md [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+   with the same result `1`. It means that one increment has been lost, and the behavior of the counter is incorrect.
+
+## Trace the invalid execution
+
+Besides showing invalid execution results, Lincheck can also provide an interleaving that leads to the error. This
+feature is accessible with the [model checking](testing-strategies.md#model-checking) testing strategy,
+which examines numerous executions with a bounded number of context switches.
+
+1. To switch the testing strategy, replace the `options` type from `StressOptions()` to `ModelCheckingOptions()`.
+The updated `CounterTest` class will look like this:
```
Comment (koshachy) [inline docs/topics/introduction.md:122]:
```suggestion
The updated `BasicCounterTest` class will look like this:
```

File: docs/topics/introduction.md [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+   ```
+
+   According to the trace, the following events have occurred:
+
+   * **T2**: The second thread starts the `inc()` operation, reading the current counter value (`value.READ: 0`) and pausing.
+   * **T1**: The first thread executes `inc()`, which returns `1`, and finishes.
+   * **T2**: The second thread resumes and increments the previously obtained counter value, incorrectly updating the
+   counter to `1`.
+
+> Get the full code [here](https://github.com/Kotlin/kotlinx-lincheck/blob/guide/src/jvm/test/org/jetbrains/kotlinx/lincheck/test/guide/BasicCounterTest.kt).
```
Comment (koshachy) [inline docs/topics/introduction.md:180]: It's better not to use the "here" word as a link (for the sake of SEO)
[Get the full code] will be better as a link

File: docs/topics/operation-arguments.md [hunk 21]
```diff
@@ -0,0 +1,135 @@
+[//]: # (title: Operation arguments)
+
+In this tutorial, you'll learn how to configure operation arguments.
```
Comment (koshachy) [inline docs/topics/operation-arguments.md:3]: I'd like to see there some info on operation arguments here

**Review by koshachy — APPROVED** [review]
Sorry, don't have enough time to look through the guide more carefully.
Some important things should be fixed (like code highlighting), others could be omitted.

Great job!

**Review by danil-pavlov — APPROVED** [review]

**Pull request merged**